File: app/backend/src/services/impactDetailsService.js
import { clamp, labelize, stateFromScore } from './formatters.js';

const areaSystems = {
  productConfiguration: ['SAP CPQ', 'SAP S/4HANA Product Master'],
  pricing: ['SAP Pricing', 'Condition Contracts'],
  charging: ['Usage Mediation', 'Rating Engine'],
  billing: ['SAP Billing', 'Subscription Billing'],
  invoicing: ['SAP S/4HANA FI-CA', 'Output Management'],
  collections: ['Collections Management', 'Dispute Management'],
  revenueRecognition: ['RAR', 'Universal Journal'],
  reporting: ['SAP Analytics Cloud', 'Data Warehouse'],
  customerService: ['SAP Service Cloud', 'Interaction Center'],
  integrations: ['Integration Suite', 'API Management']
};

function reasonFor(area, scenario, prediction, score) {
  const drivers = [];
  if (scenario.billingFrequency === 'usage-event' || scenario.pricingType === 'usage') drivers.push('usage-event charging cadence');
  if (scenario.integrationCount > 8) drivers.push(`${scenario.integrationCount} integrations`);
  if (scenario.complianceRegions.length > 2) drivers.push(`${scenario.complianceRegions.length} compliance regions`);
  if (scenario.processComplexity > 6) drivers.push(`process complexity ${scenario.processComplexity}/10`);
  if (prediction.billingComplexity > 65 && ['billing', 'charging', 'invoicing'].includes(area)) drivers.push(`predicted billing complexity ${Math.round(prediction.billingComplexity)}`);
  if (prediction.complianceRisk > 60 && area === 'revenueRecognition') drivers.push(`predicted compliance exposure ${Math.round(prediction.complianceRisk)}`);
  return drivers.length
    ? `${labelize(area)} is impacted by ${drivers.join(', ')}.`
    : `${labelize(area)} impact is primarily driven by the selected ${scenario.businessModelType} model and transaction profile.`;
}

function requiredChanges(area, scenario, prediction, score) {
  const changes = [];
  if (score >= 55) changes.push(`Update ${labelize(area)} operating design`);
  if (['charging', 'billing'].includes(area) && prediction.billingComplexity >= 55) changes.push('Validate rating, billing-event and adjustment rules');
  if (area === 'revenueRecognition') changes.push('Confirm revenue recognition treatment and close controls');
  if (area === 'integrations') changes.push('Map upstream/downstream API and data contract changes');
  if (scenario.complianceRegions.length > 1 && ['invoicing', 'reporting', 'revenueRecognition'].includes(area)) changes.push('Review regional compliance and tax reporting requirements');
  return changes.length ? changes : ['Monitor process design during implementation planning'];
}

export function buildImpactDetails(scenario, impact, predictionEnvelope) {
  const prediction = predictionEnvelope?.predictions || predictionEnvelope || {};
  const areaDetails = Object.entries(impact.o2cImpactScores || {}).map(([area, value]) => {
    const score = value.score ?? value;
    const dependencyCount = Math.max(1, Math.round((score / 25) + scenario.integrationCount / 3 + (prediction.dependencyCount || 0) / 12));
    return {
      area,
      label: labelize(area),
      score,
      severity: value.severity,
      state: stateFromScore(score),
      reason: reasonFor(area, scenario, prediction, score),
      affectedSystems: areaSystems[area] || ['SAP S/4HANA'],
      requiredChanges: requiredChanges(area, scenario, prediction, score),
      dependencyCount
    };
  });

  const riskMetrics = [
    { name: 'Implementation Risk', value: clamp(prediction.implementationRisk || impact.complianceRisk), state: stateFromScore(prediction.implementationRisk || impact.complianceRisk) },
    { name: 'Operational Effort', value: clamp(prediction.operationalEffort || impact.implementationEffort), state: stateFromScore(prediction.operationalEffort || impact.implementationEffort) },
    { name: 'Compliance Risk', value: clamp(impact.complianceRisk), state: stateFromScore(impact.complianceRisk) },
    { name: 'Delay Probability', value: clamp(impact.delayProbability), state: stateFromScore(impact.delayProbability) },
    { name: 'Integration Complexity', value: clamp(prediction.downstreamIntegrationImpact || impact.integrationImpactScores?.dataMappingComplexity), state: stateFromScore(prediction.downstreamIntegrationImpact || impact.integrationImpactScores?.dataMappingComplexity) },
    { name: 'Customer Impact', value: clamp((prediction.billingComplexity || 0) * 0.45 + (impact.delayProbability || 0) * 0.35 + scenario.processComplexity * 2), state: stateFromScore((prediction.billingComplexity || 0) * 0.45 + (impact.delayProbability || 0) * 0.35 + scenario.processComplexity * 2) }
  ];

  return {
    scenarioOverview: {
      name: scenario.name,
      businessModelType: scenario.businessModelType,
      industry: scenario.industry,
      region: scenario.region,
      billingFrequency: scenario.billingFrequency,
      contractTerm: scenario.contractTerm,
      transactionVolume: scenario.transactionVolume
    },
    areaDetails,
    riskMetrics,
    predictiveAi: {
      predictedRiskScore: prediction.implementationRisk,
      predictedRevenueImpact: prediction.revenueImpactPct,
      predictedImplementationDelay: prediction.delayProbability,
      predictedComplianceExposure: prediction.complianceRisk,
      confidenceScore: predictionEnvelope?.confidence || impact.confidenceScore,
      topContributingFactors: predictionEnvelope?.topContributingFactors || []
    },
    charts: buildScenarioCharts(scenario, impact, predictionEnvelope),
    generatedAt: new Date().toISOString()
  };
}

export function buildScenarioCharts(scenario, impact, predictionEnvelope) {
  const prediction = predictionEnvelope?.predictions || predictionEnvelope || {};
  const o2c = Object.entries(impact.o2cImpactScores || {}).map(([area, value]) => ({
    label: labelize(area),
    value: value.score ?? value,
    severity: value.severity
  }));
  const severityBreakdown = o2c.reduce((acc, item) => {
    acc[item.severity] = (acc[item.severity] || 0) + 1;
    return acc;
  }, {});
  return {
    o2cImpactByArea: o2c,
    riskEffortBreakdown: [
      { label: 'Risk', value: clamp(prediction.implementationRisk || impact.complianceRisk) },
      { label: 'Effort', value: clamp(prediction.operationalEffort || impact.implementationEffort) },
      { label: 'Compliance', value: clamp(impact.complianceRisk) },
      { label: 'Delay', value: clamp(impact.delayProbability) }
    ],
    predictionConfidence: [{ label: 'Confidence', value: clamp((predictionEnvelope?.confidence || impact.confidenceScore || 0) * 100) }],
    dependencyDistribution: [
      { label: 'Interfaces', value: impact.integrationImpactScores?.impactedInterfaces || scenario.integrationCount },
      { label: 'Model Dependencies', value: prediction.dependencyCount || 0 },
      { label: 'Compliance Regions', value: scenario.complianceRegions.length }
    ],
    severityBreakdown: Object.entries(severityBreakdown).map(([label, value]) => ({ label, value }))
  };
}

File: app/backend/src/main/java/com/bms/service/ImpactDetailsService.java
package com.bms.service;

import com.bms.util.Json;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@Service
public class ImpactDetailsService {
  private static final Map<String, List<String>> AREA_SYSTEMS = new LinkedHashMap<>() {{
    put("productConfiguration", List.of("SAP CPQ", "SAP S/4HANA Product Master")); put("pricing", List.of("SAP Pricing", "Condition Contracts"));
    put("charging", List.of("Usage Mediation", "Rating Engine")); put("billing", List.of("SAP Billing", "Subscription Billing"));
    put("invoicing", List.of("SAP S/4HANA FI-CA", "Output Management")); put("collections", List.of("Collections Management", "Dispute Management"));
    put("revenueRecognition", List.of("RAR", "Universal Journal")); put("reporting", List.of("SAP Analytics Cloud", "Data Warehouse"));
    put("customerService", List.of("SAP Service Cloud", "Interaction Center")); put("integrations", List.of("Integration Suite", "API Management"));
  }};

  public Map<String, Object> buildImpactDetails(Map<String, Object> scenario, Map<String, Object> impact, Map<String, Object> predictionEnvelope) {
    Map<String, Object> prediction = Json.map(predictionEnvelope.getOrDefault("predictions", predictionEnvelope));
    List<Map<String, Object>> areaDetails = Json.map(impact.get("o2cImpactScores")).entrySet().stream().map(entry -> {
      Map<String, Object> value = Json.map(entry.getValue());
      double score = Json.num(value.getOrDefault("score", entry.getValue()));
      int dependencyCount = Math.max(1, (int) Math.round((score / 25) + Json.num(scenario.get("integrationCount")) / 3 + Json.num(prediction.get("dependencyCount")) / 12));
      return Json.obj("area", entry.getKey(), "label", Json.labelize(entry.getKey()), "score", score, "severity", value.get("severity"),
          "state", Json.stateFromScore(score), "reason", reasonFor(entry.getKey(), scenario, prediction), "affectedSystems", AREA_SYSTEMS.getOrDefault(entry.getKey(), List.of("SAP S/4HANA")),
          "requiredChanges", requiredChanges(entry.getKey(), scenario, prediction, score), "dependencyCount", dependencyCount);
    }).toList();
    List<Map<String, Object>> riskMetrics = List.of(
        metric("Implementation Risk", Json.num(prediction.getOrDefault("implementationRisk", impact.get("complianceRisk")))),
        metric("Operational Effort", Json.num(prediction.getOrDefault("operationalEffort", impact.get("implementationEffort")))),
        metric("Compliance Risk", Json.num(impact.get("complianceRisk"))),
        metric("Delay Probability", Json.num(impact.get("delayProbability"))),
        metric("Integration Complexity", Json.num(prediction.getOrDefault("downstreamIntegrationImpact", Json.map(impact.get("integrationImpactScores")).get("dataMappingComplexity")))),
        metric("Customer Impact", Json.num(prediction.get("billingComplexity")) * 0.45 + Json.num(impact.get("delayProbability")) * 0.35 + Json.num(scenario.get("processComplexity")) * 2));
    return Json.obj(
        "scenarioOverview", Json.obj("name", scenario.get("name"), "businessModelType", scenario.get("businessModelType"), "industry", scenario.get("industry"), "region", scenario.get("region"), "billingFrequency", scenario.get("billingFrequency"), "contractTerm", scenario.get("contractTerm"), "transactionVolume", scenario.get("transactionVolume")),
        "areaDetails", areaDetails,
        "riskMetrics", riskMetrics,
        "predictiveAi", Json.obj("predictedRiskScore", prediction.get("implementationRisk"), "predictedRevenueImpact", prediction.get("revenueImpactPct"), "predictedImplementationDelay", prediction.get("delayProbability"), "predictedComplianceExposure", prediction.get("complianceRisk"), "confidenceScore", predictionEnvelope.getOrDefault("confidence", impact.get("confidenceScore")), "topContributingFactors", predictionEnvelope.getOrDefault("topContributingFactors", List.of())),
        "charts", buildScenarioCharts(scenario, impact, predictionEnvelope),
        "generatedAt", Instant.now().toString());
  }

  public Map<String, Object> buildScenarioCharts(Map<String, Object> scenario, Map<String, Object> impact, Map<String, Object> predictionEnvelope) {
    Map<String, Object> prediction = Json.map(predictionEnvelope.getOrDefault("predictions", predictionEnvelope));
    List<Map<String, Object>> o2c = Json.map(impact.get("o2cImpactScores")).entrySet().stream().map(entry -> {
      Map<String, Object> value = Json.map(entry.getValue());
      return Json.obj("label", Json.labelize(entry.getKey()), "value", value.getOrDefault("score", entry.getValue()), "severity", value.get("severity"));
    }).toList();
    Map<String, Long> severityBreakdown = o2c.stream().collect(java.util.stream.Collectors.groupingBy(row -> Json.str(row.get("severity")), LinkedHashMap::new, java.util.stream.Collectors.counting()));
    return Json.obj(
        "o2cImpactByArea", o2c,
        "riskEffortBreakdown", List.of(Json.obj("label", "Risk", "value", Json.clamp(prediction.getOrDefault("implementationRisk", impact.get("complianceRisk")))), Json.obj("label", "Effort", "value", Json.clamp(prediction.getOrDefault("operationalEffort", impact.get("implementationEffort")))), Json.obj("label", "Compliance", "value", Json.clamp(impact.get("complianceRisk"))), Json.obj("label", "Delay", "value", Json.clamp(impact.get("delayProbability")))),
        "predictionConfidence", List.of(Json.obj("label", "Confidence", "value", Json.clamp(Json.num(predictionEnvelope.getOrDefault("confidence", impact.get("confidenceScore"))) * 100))),
        "dependencyDistribution", List.of(Json.obj("label", "Interfaces", "value", Json.map(impact.get("integrationImpactScores")).getOrDefault("impactedInterfaces", scenario.get("integrationCount"))), Json.obj("label", "Model Dependencies", "value", prediction.getOrDefault("dependencyCount", 0)), Json.obj("label", "Compliance Regions", "value", Json.array(scenario.get("complianceRegions")).size())),
        "severityBreakdown", severityBreakdown.entrySet().stream().map(e -> Json.obj("label", e.getKey(), "value", e.getValue())).toList());
  }

  private Map<String, Object> metric(String name, double value) {
    int clamped = Json.clamp(value);
    return Json.obj("name", name, "value", clamped, "state", Json.stateFromScore(clamped));
  }

  private String reasonFor(String area, Map<String, Object> scenario, Map<String, Object> prediction) {
    java.util.ArrayList<String> drivers = new java.util.ArrayList<>();
    if ("usage-event".equals(scenario.get("billingFrequency")) || "usage".equals(scenario.get("pricingType"))) drivers.add("usage-event charging cadence");
    if (Json.num(scenario.get("integrationCount")) > 8) drivers.add(Json.integer(scenario.get("integrationCount")) + " integrations");
    if (Json.array(scenario.get("complianceRegions")).size() > 2) drivers.add(Json.array(scenario.get("complianceRegions")).size() + " compliance regions");
    if (Json.num(scenario.get("processComplexity")) > 6) drivers.add("process complexity " + Json.integer(scenario.get("processComplexity")) + "/10");
    if (Json.num(prediction.get("billingComplexity")) > 65 && List.of("billing", "charging", "invoicing").contains(area)) drivers.add("predicted billing complexity " + Math.round(Json.num(prediction.get("billingComplexity"))));
    if (Json.num(prediction.get("complianceRisk")) > 60 && "revenueRecognition".equals(area)) drivers.add("predicted compliance exposure " + Math.round(Json.num(prediction.get("complianceRisk"))));
    return drivers.isEmpty() ? Json.labelize(area) + " impact is primarily driven by the selected " + scenario.get("businessModelType") + " model and transaction profile." : Json.labelize(area) + " is impacted by " + String.join(", ", drivers) + ".";
  }

  private List<String> requiredChanges(String area, Map<String, Object> scenario, Map<String, Object> prediction, double score) {
    java.util.ArrayList<String> changes = new java.util.ArrayList<>();
    if (score >= 55) changes.add("Update " + Json.labelize(area) + " operating design");
    if (List.of("charging", "billing").contains(area) && Json.num(prediction.get("billingComplexity")) >= 55) changes.add("Validate rating, billing-event and adjustment rules");
    if ("revenueRecognition".equals(area)) changes.add("Confirm revenue recognition treatment and close controls");
    if ("integrations".equals(area)) changes.add("Map upstream/downstream API and data contract changes");
    if (Json.array(scenario.get("complianceRegions")).size() > 1 && List.of("invoicing", "reporting", "revenueRecognition").contains(area)) changes.add("Review regional compliance and tax reporting requirements");
    return changes.isEmpty() ? List.of("Monitor process design during implementation planning") : changes;
  }
}
